### PR TITLE
[Feature] Add DOTAMeanAP metric

### PR DESCRIPTION
**Pull request by YanxingLiu**
## Motivation

Add dota metric for rotated boxes evaluation

The PR link that use MMEval.DOTAMeanAP in mmrotate: https://github.com/open-mmlab/mmrotate/pull/669

## Modification

Add dota metric for rotated boxes evaluation

## Use cases (Optional)
```python
import numpy as np
from mmeval import DOTAMetric

num_classes = 15
dota_metric = DOTAMetric(num_classes=15)

def _gen_bboxes(num_bboxes, img_w=256, img_h=256):
    x = np.random.rand(num_bboxes, ) * img_w
    y = np.random.rand(num_bboxes, ) * img_h
    w = np.random.rand(num_bboxes, ) * (img_w - x)
    h = np.random.rand(num_bboxes, ) * (img_h - y)
    a = np.random.rand(num_bboxes, ) * np.pi / 2
    return np.stack([x, y, w, h, a], axis=1)

prediction = {
    'bboxes': _gen_bboxes(10),
    'scores': np.random.rand(10, ),
    'labels': np.random.randint(0, num_classes, size=(10, ))
}
groundtruth = {
    'bboxes': _gen_bboxes(10),
    'labels': np.random.randint(0, num_classes, size=(10, )),
    'bboxes_ignore': _gen_bboxes(5),
    'labels_ignore': np.random.randint(0, num_classes, size=(5, ))
}
dota_metric(predictions=[prediction, ], groundtruths=[groundtruth, ]) # doctest: +ELLIPSIS # noqa: E501
```

### CODE REVIEW

File: mmeval/metrics/dota_map.py [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+    """Filter the rotated bboxes with an area range.
+
+    Args:
+        bboxes (numpy.ndarray): The bboxes with shape (n, 5) in 'xywha' format.
+        min_area (Optional[float]): The minimum area. If None, does not filter
+            the minimum area.
+        max_area (Optional[float]): The maximum area. If None, does not filter
+            the maximum area.
+    Returns:
+        numpy.ndarray: A mask of ``bboxes`` identify which bbox are filtered.
```
Comment (ice-tong) [inline mmeval/metrics/dota_map.py:23]: Need a blank line before 'Returns'

File: mmeval/metrics/dota_map.py [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+    if max_area is not None:
+        area_mask &= (bboxes_area < max_area)
+    return area_mask
+
+
+class DOTAMeanAP(VOCMeanAP):
+    """DOTA evaluation metric.
+    This metric computes the DOTA mAP (mean Average Precision) with the given
+    IoU thresholds and scale ranges.
+    Args:
```
Comment (ice-tong) [inline mmeval/metrics/dota_map.py:38]: About the docstring:

1. Need add blank line for each section or block. Please reference to the docstring of `VOCMeanAP`.
2. We should add a brief introduction of DOTA dataset and add the official link: https://captain-whu.github.io/DOTA/dataset.html

File: mmeval/metrics/dota_map.py [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+            Defaults to '11points'.
+        nproc (int): Processes used for computing TP and FP. If nproc
+            is less than or equal to 1, multiprocessing will not be used.
+            Defaults to 4.
+        drop_class_ap (bool): Whether to drop the class without ground truth
+            when calculating the average precision for each class.
+        classwise (bool): Whether to return the computed results of each
+            class. Defaults to False.
+        **kwargs: Keyword parameters passed to :class:`BaseMetric`.
+    Examples:
```
Comment (ice-tong) [inline mmeval/metrics/dota_map.py:60]: Need a blank line before 'Examples'

File: mmeval/metrics/utils/bbox_iou_rotated.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    """
+    bboxes_w = bboxes[..., 2]
+    bboxes_h = bboxes[..., 3]
+    areas = bboxes_w * bboxes_h
+    return areas
+
+
+def bbox_iou_rotated(bboxes1: np.ndarray,
+                     bboxes2: np.ndarray,
+                     clockwise: bool = True) -> np.ndarray:
```
Comment (ice-tong) [inline mmeval/metrics/utils/bbox_iou_rotated.py:48]: We need try to import mmcv and priority use the bbox iou rotated op in mmcv~

File: mmeval/metrics/dota_map.py [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+        ...     'bboxes': _gen_bboxes(10),
+        ...     'labels': np.random.randint(0, num_classes, size=(10, )),
+        ...     'bboxes_ignore': _gen_bboxes(5),
+        ...     'labels_ignore': np.random.randint(0, num_classes, size=(5, ))
+        ... }
+        >>> dota_metric(predictions=[prediction, ], groundtruths=[groundtruth, ])  # doctest: +ELLIPSIS  # noqa: E501
+        {'mAP@0.5': ..., 'mAP': ...}
+    """
+
+    def __init__(self, eval_mode: str = '11points', **kwargs) -> None:
```
Comment (ice-tong) [inline mmeval/metrics/dota_map.py:104]: It is better to write out all the kwargs corresponds to the docstring~

File: mmeval/metrics/dota_map.py [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+from .utils.bbox_overlaps_rotated import calculate_bboxes_area_rotated
+
+logger = logging.getLogger(__name__)
+try:
+    # we prefer to use bbox_iou_rotated in mmcv
+    from mmcv.ops import box_iou_rotated as calculate_overlaps_rotated
+except Exception as e:  # noqa F841
+    logger.info(
+        'mmcv is not installed, using iou calculating method with OpenCV.')
+    from .utils.bbox_overlaps_rotated import calculate_overlaps_rotated
```
Comment (ice-tong) [inline mmeval/metrics/dota_map.py:17]: This should moved to `mmeval/metrics/utils/bbox_overlaps_rotated.py`, and log the warning only when `calculate_overlaps_rotated` is used.

File: mmeval/metrics/dota_map.py [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+
+        image_tp_list, image_fp_list = tuple(zip(*tpfp_list))
+        sorted_indices = np.argsort(-np.vstack(class_preds)[:, -1])
+        tp = np.concatenate(image_tp_list, axis=2)[..., sorted_indices]
+        fp = np.concatenate(image_fp_list, axis=2)[..., sorted_indices]
+        num_gts = np.zeros((self.num_iou, self.num_scale), dtype=int)
+        for idx, (min_area, max_area) in enumerate(self._area_ranges):
+            area_mask = filter_by_bboxes_area_rotated(
+                np.vstack(class_gts), min_area, max_area)
+            num_gts[:, idx] = np.sum(area_mask)
```
Comment (ice-tong) [inline mmeval/metrics/dota_map.py:292]: The `filter_by_bboxes_area_rotated` seems should be a class method to avoid the overwrite of `calculate_class_tpfp`?

File: mmeval/metrics/dota_map.py (hunk 6 above)
Comment (ice-tong) [inline mmeval/metrics/dota_map.py:17]:
```suggestion
logger = logging.getLogger(__name__)

try:
    # we prefer to use `bbox_iou_rotated` in mmcv
    from mmcv.ops import box_iou_rotated as calculate_overlaps_rotated
except Exception as e:  # noqa F841
    from .utils.bbox_overlaps_rotated import calculate_overlaps_rotated
    logger.debug(
        'mmcv is not installed, calculating IoU of rotated bbox with OpenCV.')
```

File: mmeval/metrics/dota_map.py [hunk 8]
```diff
@@ -0,0 +1,294 @@
...
+            is less than or equal to 1, multiprocessing will not be used.
+            Defaults to 4.
+        drop_class_ap (bool): Whether to drop the class without ground truth
+            when calculating the average precision for each class.
+        classwise (bool): Whether to return the computed results of each
+            class. Defaults to False.
+        **kwargs: Keyword parameters passed to :class:`BaseMetric`.
+
+    Examples:
+        >>> import numpy as np
```
Comment (ice-tong) [inline mmeval/metrics/dota_map.py:76]:
```suggestion
    Examples:

        >>> import numpy as np
```

File: mmeval/metrics/end_point_error.py [hunk 9]
```diff
@@ -149,13 +149,13 @@ def end_point_error_map(
...
-    def end_point_error_map(
+    @dispatch
+    def end_point_error_map(  # noqa: F811
         self,
         prediction: 'oneflow.Tensor',
         label: 'oneflow.Tensor',
-        valid_mask: Optional['oneflow.Tensor'] = None
-    ) -> Tuple[np.ndarray, int]:
+        valid_mask: Optional['oneflow.Tensor'] = None) -> Tuple[np.ndarray,
+                                                                int]:
```
Comment (ice-tong) [inline mmeval/metrics/end_point_error.py:158]: Unnecessary changes~ This seems a problem related `pre-commit`.

File: mmeval/metrics/voc_map.py [hunk 10]
```diff
@@ -560,3 +560,22 @@ def _aggregate_results(self, results_per_class: List[dict]) -> dict:
             eval_results[key] = np.array(ap_per_ious).mean().item()
 
         return eval_results
+
+    def filter_by_bboxes_area(self, bboxes: np.ndarray,
+                              min_area: Optional[float],
+                              max_area: Optional[float]):
```
Comment (ice-tong) [inline mmeval/metrics/voc_map.py:566]: Suggest names: `_filter_by_bboxes_area `

File: mmeval/metrics/end_point_error.py [hunk 11]
```diff
@@ -149,13 +149,13 @@ def end_point_error_map(
         epe = epe_map.reshape(-1)[val]
         return epe.mean().cpu().numpy(), int(val.sum())
 
-    @dispatch  # noqa: F811
-    def end_point_error_map(
+    @dispatch
+    def end_point_error_map(  # noqa: F811
```
Comment (ice-tong) [inline mmeval/metrics/end_point_error.py:153]:
```suggestion
    @dispatch  # noqa: F811
    def end_point_error_map(
```

File: mmeval/metrics/end_point_error.py (hunk 9 above)
Comment (ice-tong) [inline mmeval/metrics/end_point_error.py:158]:
```suggestion
        valid_mask: Optional['oneflow.Tensor'] = None
    ) -> Tuple[np.ndarray, int]:
```

File: mmeval/metrics/utils/bbox_overlaps_rotated.py [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+
+
+def qbox_to_rbox(boxes: np.ndarray) -> np.ndarray:
+    """Convert quadrilateral boxes to rotated boxes.
+
+    Args:
+        boxes (np.ndarray): Quadrilateral box tensor with shape of (..., 8).
+
+    Returns:
+        np.ndarray: Rotated box tensor with shape of (..., 5).
```
Comment (ice-tong) [inline mmeval/metrics/utils/bbox_overlaps_rotated.py:13]:
```suggestion
        np.ndarray: Rotated box array with shape of (..., 5).
```

File: mmeval/metrics/dota_map.py [hunk 13]
```diff
@@ -0,0 +1,305 @@
...
+        on the ``self.predict_box_type``. If ``self.predict_box_type`` is
+        'rbox', the box shape should be (N, 5) which represents the (x,y,w,h,
+        angle), otherwise the box shape should be (N, 8) which represents the
+        (x1,y1,x2,y2,x3,y3,x4,y4).
+
+        Args:
+            predictions (Sequence[Dict]):  A sequence of dict. Each dict
+                representing a detection result for an image, with the
+                following keys:
+                - bboxes (numpy.ndarray): Shape (N, 5) or shape (N, 8).
```
Comment (ice-tong) [inline mmeval/metrics/dota_map.py:146]:
```suggestion
                following keys:
                
                - bboxes (numpy.ndarray): Shape (N, 5) or shape (N, 8).
```

File: mmeval/metrics/dota_map.py [hunk 14]
```diff
@@ -0,0 +1,305 @@
...
+                  bounding bboxes of this image. The box format is depend on
+                  ``self.predict_box_type``. Details in upper note.
+                - scores (numpy.ndarray): Shape (N, ), the predicted scores
+                  of bounding boxes.
+                - labels (numpy.ndarray): Shape (N, ), the predicted labels
+                  of bounding boxes.
+            groundtruths (Sequence[Dict]):  A sequence of dict. Each dict
+                represents a groundtruths for an image, with the following
+                keys:
+                - bboxes (numpy.ndarray): Shape (M, 5) or shape (M, 8), the
```
Comment (ice-tong) [inline mmeval/metrics/dota_map.py:156]:
```suggestion
                  of bounding boxes.

            groundtruths (Sequence[Dict]):  A sequence of dict. Each dict
                represents a groundtruths for an image, with the following
                keys:

                - bboxes (numpy.ndarray): Shape (M, 5) or shape (M, 8), the
```

File: mmeval/metrics/utils/bbox_overlaps_rotated.py [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+    rboxes = np.array(rboxes)
+    return rboxes.reshape(*original_shape, 5)
+
+
+def le90_to_oc(bboxes: np.ndarray):
+    """convert bboxes with le90 angle version to OpenCV angle version.
+
+    Args:
+        bboxes (np.ndarray): The shape of bboxes should be [N,5],
+        the format is [x,y,w,h,angle]
```
Comment (ice-tong) [inline mmeval/metrics/utils/bbox_overlaps_rotated.py:30]:
```suggestion
        bboxes (np.ndarray): The shape of bboxes should be (N, 5),
        the format is 'xywha'.
```

File: mmeval/metrics/utils/bbox_overlaps_rotated.py [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+
+    Returns:
+        np.ndarray: An numpy.ndarray with the same shape of input.
+    """
+    assert bboxes.shape[1] == 5, 'The boxes shape should be [N,5]'
```
Comment (ice-tong) [inline mmeval/metrics/utils/bbox_overlaps_rotated.py:35]:
```suggestion
    assert bboxes.shape[1] == 5, 'The boxes shape should be (N, 5)'
```

File: mmeval/metrics/utils/bbox_overlaps_rotated.py [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+    ret_bboxes[:, 4] = ret_bboxes[:, 4] * 180.0 / np.pi
+    return ret_bboxes
+
+
+def calculate_bboxes_area_rotated(bboxes: np.ndarray) -> np.ndarray:
+    """Calculate area of rotated bounding boxes.
+
+    Args:
+        bboxes (np.ndarray): The bboxes with shape (n, 5) or (5, )
+        in 'xywha'format.
```
Comment (ice-tong) [inline mmeval/metrics/utils/bbox_overlaps_rotated.py:56]:
```suggestion
        in 'xywha' format.
```

File: mmeval/metrics/utils/bbox_overlaps_rotated.py [hunk 18]
```diff
@@ -0,0 +1,115 @@
...
+def calculate_overlaps_rotated(bboxes1: np.ndarray,
+                               bboxes2: np.ndarray,
+                               clockwise: bool = True) -> np.ndarray:
+    """Calculate the overlap between each rotated bbox of bboxes1 and bboxes2.
+
+    Args:
+        bboxes1 (np.ndarray): The bboxes with shape (n, 5) in 'xywha' format.
+        bboxes2 (np.ndarray): The bboxes with shape (k, 5) in 'xywha' format.
+        clockwise (bool, optional): flag indicating whether the positive
+        angular orientation is clockwise. Defaults to True.
```
Comment (ice-tong) [inline mmeval/metrics/utils/bbox_overlaps_rotated.py:76]:
```suggestion
        clockwise (bool, optional): flag indicating whether the positive
            angular orientation is clockwise. Defaults to True.
```

File: mmeval/metrics/dota_map.py [hunk 19]
```diff
@@ -0,0 +1,308 @@
...
+        on the ``self.predict_box_type``. If ``self.predict_box_type`` is
+        'rbox', the box shape should be (N, 5) which represents the (x,y,w,h,
+        angle), otherwise the box shape should be (N, 8) which represents the
+        (x1,y1,x2,y2,x3,y3,x4,y4).
+
+        Args:
+            predictions (Sequence[Dict]):  A sequence of dict. Each dict
+                representing a detection result for an image, with the
+                following keys:
+
```
Comment (ice-tong) [inline mmeval/metrics/dota_map.py:146]:
```suggestion

```

**Comment by ice-tong** [comment]
Please add the DOTAMeanAP API reference in `docs/zh_cn/api/metrics.rst` and `docs/en/api/metrics.rst`.

**Comment by ice-tong** [comment]
Please @zytx121 have a look~

File: mmeval/metrics/dota_map.py [hunk 20]
```diff
@@ -0,0 +1,307 @@
...
+        self.predict_box_type = predict_box_type
+
+    def add(self, predictions: Sequence[Dict], groundtruths: Sequence[Dict]) -> None:  # type: ignore # yapf: disable # noqa: E501
+        """Add the intermediate results to ``self._results``.
+
+        Note: The box shape of ``predictions`` and ``groundtruths`` is depends
+        on the ``self.predict_box_type``. If ``self.predict_box_type`` is
+        'rbox', the box shape should be (N, 5) which represents the (x,y,w,h,
+        angle), otherwise the box shape should be (N, 8) which represents the
+        (x1,y1,x2,y2,x3,y3,x4,y4).
```
Comment (ice-tong) [inline mmeval/metrics/dota_map.py:140]:
```suggestion
        'rbox', the box shape should be (N, 5) which represents the (x, y, w, h,
        angle), otherwise the box shape should be (N, 8) which represents the
        (x1, y1, x2, y2, x3, y3, x4, y4).
```

File: mmeval/metrics/dota_map.py [hunk 21]
```diff
@@ -0,0 +1,307 @@
...
+                               min_area: Optional[float],
+                               max_area: Optional[float]):
+        """Filter the bboxes with an area range.
+
+        Args:
+            bboxes (numpy.ndarray): The bboxes with shape (n, 5) in 'xywha'
+                format.
+            min_area (Optional[float]): The minimum area. If None, does not
+                filter the minimum area.
+            max_area (_type_): The maximum area. If None, does not filter
```
Comment (ice-tong) [inline mmeval/metrics/dota_map.py:300]:
```suggestion
            max_area (Optional[float]): The maximum area. If None, does not filter
```

File: mmeval/metrics/voc_map.py [hunk 22]
```diff
@@ -560,3 +560,22 @@ def _aggregate_results(self, results_per_class: List[dict]) -> dict:
...
+                               min_area: Optional[float],
+                               max_area: Optional[float]):
+        """Filter the bboxes with an area range.
+
+        Args:
+            bboxes (numpy.ndarray): The bboxes with shape (n, 4) in
+                'xyxy' format.
+            min_area (Optional[float]): The minimum area. If None, does
+                not filter the minimum area.
+            max_area (_type_): The maximum area. If None, does not filter
```
Comment (ice-tong) [inline mmeval/metrics/voc_map.py:574]:
```suggestion
            max_area (Optional[float]): The maximum area. If None, does not filter
```

**Review by ice-tong — APPROVED** [review]
LGTM with 3 comments

**Comment by ice-tong** [comment]
Please add the DOTAMeanAP API reference in docs/zh_cn/api/metrics.rst and docs/en/api/metrics.rst.

File: mmeval/metrics/dota_map.py (hunk 20 above)
Comment (ice-tong) [inline mmeval/metrics/dota_map.py:140]:
```suggestion
        Note:
            The box shape of ``predictions`` and ``groundtruths`` is depends
            on the ``self.predict_box_type``. If ``self.predict_box_type`` is
            'rbox', the box shape should be (N, 5) which represents the (x, y, w, h,
            angle), otherwise the box shape should be (N, 8) which represents the
            (x1, y1, x2, y2, x3, y3, x4, y4).
```

**Review by zytx121 — APPROVED** [review]
LGTM

File: mmeval/metrics/dota_map.py [hunk 23]
```diff
@@ -0,0 +1,307 @@
...
+            iou_thrs=iou_thrs,
+            scale_ranges=scale_ranges,
+            num_classes=num_classes,
+            eval_mode=eval_mode,
+            use_legacy_coordinate=False,
+            nproc=nproc,
+            drop_class_ap=drop_class_ap,
+            classwise=classwise,
+            **kwargs)
+        self.predict_box_type = predict_box_type
```
Comment (ice-tong) [inline mmeval/metrics/dota_map.py:131]: The `self.predict_box_type ` seems never used~

File: mmeval/metrics/utils/bbox_overlaps_rotated.py [hunk 24]
```diff
@@ -0,0 +1,115 @@
...
+    rboxes = np.array(rboxes)
+    return rboxes.reshape(*original_shape, 5)
+
+
+def le90_to_oc(bboxes: np.ndarray):
+    """convert bboxes with le90 angle version to OpenCV angle version.
+
+    Args:
+        bboxes (np.ndarray): The shape of bboxes should be (N, 5),
+        the format is 'xywha'.
```
Comment (zhouzaida) [inline mmeval/metrics/utils/bbox_overlaps_rotated.py:30]:
```suggestion
        bboxes (np.ndarray): The shape of bboxes should be (N, 5),
            the format is 'xywha'.
```

File: mmeval/metrics/utils/bbox_overlaps_rotated.py [hunk 25]
```diff
@@ -0,0 +1,135 @@
...
+def calculate_overlaps_rotated(bboxes1: np.ndarray,
+                               bboxes2: np.ndarray,
+                               clockwise: bool = True) -> np.ndarray:
+    """Calculate the overlap between each rotated bbox of bboxes1 and bboxes2.
+
+       Modified from: https://github.com/yangxue0827/RotationDetection/blob/
+       45c708a34348e38040786ba85b288bc147786452/alpharotate/libs/
+       utils/iou_rotate.py#L39
+
+    Args:
```
Comment (ice-tong) [inline mmeval/metrics/utils/bbox_overlaps_rotated.py:85]:
```suggestion
    """Calculate the overlap between each rotated bbox of bboxes1 and bboxes2.

    Modified from: https://github.com/yangxue0827/RotationDetection/blob/
    45c708a34348e38040786ba85b288bc147786452/alpharotate/libs/
    utils/iou_rotate.py#L39

    Args:
```

File: mmeval/metrics/utils/bbox_overlaps_rotated.py [hunk 26]
```diff
@@ -0,0 +1,135 @@
...
+
+    Returns:
+        np.ndarray: An numpy.ndarray with the same shape of input.
+    """
+    assert bboxes.ndim == 2 and bboxes.shape[-1] == 5, \
+        'The boxes shape should be (N, 5)'
+
+    # a mask to indicate if input angles belong to [-pi/2,0)
+    mask = bboxes[:, 4] <= 0.0
+    # convert the boxes whose angle is in range of [-pi/2,0)
```
Comment (ice-tong) [inline mmeval/metrics/utils/bbox_overlaps_rotated.py:43]:
```suggestion
    # a mask to indicate if input angles belong to [-pi/2, 0)
    mask = bboxes[:, 4] <= 0.0
    # convert the boxes whose angle is in range of [-pi/2, 0)
```

File: mmeval/metrics/utils/bbox_overlaps_rotated.py [hunk 27]
```diff
@@ -0,0 +1,135 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import cv2
```
Comment (zhouzaida) [inline mmeval/metrics/utils/bbox_overlaps_rotated.py:2]: `opencv-python` is not included in `requirements/runtime.txt` so importing cv2 will cause an error.

File: mmeval/metrics/utils/bbox_overlaps_rotated.py [hunk 28]
```diff
@@ -0,0 +1,135 @@
...
+    ret_bboxes[mask] = temp
+    return ret_bboxes
+
+
+def calculate_bboxes_area_rotated(bboxes: np.ndarray) -> np.ndarray:
+    """Calculate area of rotated bounding boxes.
+
+    Args:
+        bboxes (np.ndarray): The bboxes with shape (n, 5) or (5, )
+        in 'xywha' format.
```
Comment (zhouzaida) [inline mmeval/metrics/utils/bbox_overlaps_rotated.py:65]:
```suggestion
            in 'xywha' format.
```

File: mmeval/metrics/dota_map.py [hunk 29]
```diff
@@ -0,0 +1,303 @@
...
+    from torch import Tensor
+    HAS_MMCV = True
+except Exception as e:  # noqa F841
+    from .utils.bbox_overlaps_rotated import calculate_overlaps_rotated
+    HAS_MMCV = False
+    logger.debug(
+        'mmcv is not installed, calculating IoU of rotated bbox with OpenCV.')
+
+
+def filter_by_bboxes_area_rotated(bboxes: np.ndarray,
```
Comment (zhouzaida) [inline mmeval/metrics/dota_map.py:24]: This function has been already defined in `voc_map.py`.

File: mmeval/metrics/dota_map.py [hunk 30]
```diff
@@ -0,0 +1,303 @@
...
+                            fp[iou_thr_idx, area_idx, pred_bbox_idx] = 1
+                    else:
+                        area_mask = filter_by_bboxes_area_rotated(
+                            pred_bboxes[pred_bbox_idx, :5], min_area, max_area)
+                        if area_mask:
+                            fp[iou_thr_idx, area_idx, pred_bbox_idx] = 1
+
+        return tp, fp
+
+    def _filter_by_bboxes_area(self, bboxes: np.ndarray,
```
Comment (zhouzaida) [inline mmeval/metrics/dota_map.py:286]: This method is not called by `_calculate_image_tpfp`.

File: mmeval/metrics/utils/bbox_overlaps_rotated.py (hunk 27 above)
Comment (ice-tong) [reply mmeval/metrics/utils/bbox_overlaps_rotated.py]: Fixed by use try_import

File: mmeval/metrics/dota_map.py (hunk 29 above)
Comment (ice-tong) [reply mmeval/metrics/dota_map.py:24]: The similar function in voc_map.py is named `filter_by_bboxes_area`~

File: mmeval/metrics/dota_map.py (hunk 30 above)
Comment (ice-tong) [reply mmeval/metrics/dota_map.py:286]: This method is called by `VOCMeanAP.calculate_class_tpfp` to avoid overwrite a lot of code in `calculate_class_tpfp`

File: mmeval/metrics/dota_map.py [hunk 31]
```diff
@@ -0,0 +1,303 @@
...
+        # Sort all pred bbox in descending order by scores.
+        sorted_indices = np.argsort(-pred_bboxes[:, -1])
+
+        # Step 5. Count the `tp` and `fp` of each iou threshold and area range.
+        for iou_thr_idx, iou_thr in enumerate(iou_thrs):
+            for area_idx, (min_area, max_area) in enumerate(area_ranges):
+                # The flags that gt bboxes have been matched.
+                gt_covered_flags = np.zeros(all_gt_bboxes.shape[0], dtype=bool)
+                # The flags that gt bboxes out of area range.
+                gt_area_mask = filter_by_bboxes_area_rotated(
```
Comment (zhouzaida) [inline mmeval/metrics/dota_map.py:258]:
```suggestion
                gt_area_mask = self._filter_by_bboxes_area(
```

File: mmeval/metrics/dota_map.py [hunk 32]
```diff
@@ -0,0 +1,303 @@
...
+                        if ignore_gt_area_flags[matched_gt_idx]:
+                            continue
+                        if not gt_covered_flags[matched_gt_idx]:
+                            tp[iou_thr_idx, area_idx, pred_bbox_idx] = 1
+                            gt_covered_flags[matched_gt_idx] = True
+                        else:
+                            # This gt bbox has been matched and counted as fp.
+                            fp[iou_thr_idx, area_idx, pred_bbox_idx] = 1
+                    else:
+                        area_mask = filter_by_bboxes_area_rotated(
```
Comment (zhouzaida) [inline mmeval/metrics/dota_map.py:279]:
```suggestion
                        area_mask = self._filter_by_bboxes_area(
```

File: mmeval/metrics/dota_map.py [hunk 33]
```diff
@@ -0,0 +1,303 @@
...
+                                  min_area: Optional[float],
+                                  max_area: Optional[float]):
+    """Filter the rotated bboxes with an area range.
+
+    Args:
+        bboxes (numpy.ndarray): The bboxes with shape (n, 5) in 'xywha' format.
+        min_area (Optional[float]): The minimum area. If None, does not filter
+            the minimum area.
+        max_area (Optional[float]): The maximum area. If None, does not filter
+            the maximum area.
```
Comment (zhouzaida) [inline mmeval/metrics/dota_map.py:34]:
```suggestion
        min_area (float, optional): The minimum area. If None, do not filter
            the minimum area.
        max_area (float, optional): The maximum area. If None, do not filter
            the maximum area.
```

File: mmeval/metrics/dota_map.py [hunk 34]
```diff
@@ -0,0 +1,303 @@
...
+            bboxes (numpy.ndarray): The bboxes with shape (n, 5) in 'xywha'
+                format.
+            min_area (Optional[float]): The minimum area. If None, does not
+                filter the minimum area.
+            max_area (Optional[float]): The maximum area. If None, does not
+                filter the maximum area.
+
+        Returns:
+            numpy.ndarray: A mask of ``bboxes`` identify which bbox are
+                filtered.
```
Comment (zhouzaida) [inline mmeval/metrics/dota_map.py:301]:
```suggestion
            numpy.ndarray: A mask of ``bboxes`` identify which bbox are
            filtered.
```

File: mmeval/metrics/utils/bbox_overlaps_rotated.py [hunk 35]
```diff
@@ -0,0 +1,150 @@
...
+
+    Args:
+        boxes (np.ndarray): Quadrilateral box tensor with shape of (N, 8).
+
+    Returns:
+        np.ndarray: Rotated box array with shape of (N, 5).
+    """
+    if cv2 is None:
+        raise ImportError('To use `qbox_to_rbox` function, '
+                          'please install opecnv-python first.')
```
Comment (zhouzaida) [inline mmeval/metrics/utils/bbox_overlaps_rotated.py:24]:
```suggestion
                          'please install opencv-python first.')
```

File: mmeval/metrics/utils/bbox_overlaps_rotated.py [hunk 36]
```diff
@@ -0,0 +1,150 @@
...
+        bboxes2 (np.ndarray): The bboxes with shape (k, 5) in 'xywha' format.
+        clockwise (bool, optional): flag indicating whether the positive
+            angular orientation is clockwise. Defaults to True.
+
+    Returns:
+        np.ndarray: IoUs with shape (n, k).
+    """
+    if cv2 is None:
+        raise ImportError('To use `calculate_overlaps_rotated` function, '
+                          'please install opecnv-python first.')
```
Comment (zhouzaida) [inline mmeval/metrics/utils/bbox_overlaps_rotated.py:107]:
```suggestion
                          'please install opencv-python first.')
```

File: mmeval/metrics/voc_map.py [hunk 37]
```diff
@@ -560,3 +560,23 @@ def _aggregate_results(self, results_per_class: List[dict]) -> dict:
...
+            bboxes (numpy.ndarray): The bboxes with shape (n, 4) in
+                'xyxy' format.
+            min_area (Optional[float]): The minimum area. If None, does
+                not filter the minimum area.
+            max_area (Optional[float]): The maximum area. If None, does
+                not filter the maximum area.
+
+        Returns:
+            numpy.ndarray: A mask of ``bboxes`` identify which bbox
+                are filtered.
```
Comment (zhouzaida) [inline mmeval/metrics/voc_map.py:579]:
```suggestion
            numpy.ndarray: A mask of ``bboxes`` identify which bbox
            are filtered.
```

File: tests/test_metrics/test_dota_map.py [hunk 38]
```diff
@@ -0,0 +1,71 @@
...
+            _gen_prediction(num_classes=num_classes) for _ in range(10)
+        ],
+        groundtruths=[
+            _gen_groundtruth(num_classes=num_classes) for _ in range(10)
+        ],
+    )
+    assert isinstance(metric_results, dict)
+    assert 'mAP' in metric_results
+
+
```
Comment (zhouzaida) [inline tests/test_metrics/test_dota_map.py:69]: Please refer to https://github.com/open-mmlab/mmeval/blob/e635ba6338105d92f053b46320f3b1ca340ef9b4/tests/test_metrics/test_voc_map.py#L133 to validate the accuracy of `DOTAMeanAP`.

Comment (YanxingLiu) [reply tests/test_metrics/test_dota_map.py:69]: I have added a simple test_metric_accurate in test_dota_map.py. More accurate test  results are in https://github.com/open-mmlab/mmrotate/pull/669. Thank you for your reviewing.

**Review by zhouzaida — APPROVED** [review]
